contracts/nexus-gateway/src/contract/execute.rs

```diff
@@ -86,23 +26,66 @@ pub fn route_to_router(
     Ok(Response::new().add_messages(router.route(msgs)))
 }
 
-pub fn route_to_nexus(
+pub fn route_message_with_token_to_nexus(
     storage: &mut dyn Storage,
+    querier: QuerierWrapper,
+    info: MessageInfo,
+    msg: router_api::Message,
+) -> Result<Response<nexus::execute::Message>> {
+    let token = match info.funds.as_slice() {
```

Comment:
Use `info.funds.pop()` + a length check instead to avoid cloning the token? Could make it into a helper for better readability.

Reply:
Same here, I don't like a mutable reference to the MessageInfo.

Reply:
I had exactly the same train of thought for an optimization, but `mut MessageInfo` is a big no for me too